heron/dsl/src/java/com/twitter/heron/dsl/impl/operators/JoinOperator.java

```diff
@@ -0,0 +1,135 @@
+ * It embeds the logic of the type of join(outer, left, inner) which it takes in as
+ * a config parameter. Also taken as parameters are which source is left and right.
+ * This is needed for the semantics of outer/left/inner joins.
+ */
+public class JoinOperator<K, V1, V2, VR> extends DslWindowOperator {
+  private static final long serialVersionUID = 4875450390444745407L;
+  public static final String LEFT_COMPONENT_NAME = "_dsl_joinbolt_left_component_name_";
+  public static final String RIGHT_COMPONENT_NAME = "_dsl_joinbolt_right_component_name_";
+
+  public enum JoinType {
+    INNER,
+    LEFT,
+    OUTER;
+  }
+
+  private JoinType joinType;
+  // The source component that represent the left join component
+  private String leftComponent;
+  // The source component that represent the right join component
+  private String rightComponent;
+  // The user supplied join function
+  private SerializableBiFunction<? super V1, ? super V2, ? extends VR> joinFn;
+  private OutputCollector collector;
+
+  public JoinOperator(JoinType joinType, String leftComponent, String rightComponent,
+                      SerializableBiFunction<? super V1, ? super V2, ? extends VR> joinFn) {
+    this.joinType = joinType;
+    this.leftComponent = leftComponent;
+    this.rightComponent = rightComponent;
+    this.joinFn = joinFn;
+  }
+
+  @SuppressWarnings("rawtypes")
+  @Override
+  public void prepare(Map map, TopologyContext topologyContext, OutputCollector outputCollector) {
+    collector = outputCollector;
+  }
+
+  @Override
+  public Map<String, Object> getComponentConfiguration() {
+    Map<String, Object> cfg = super.getComponentConfiguration();
+    cfg.put(LEFT_COMPONENT_NAME, leftComponent);
+    cfg.put(RIGHT_COMPONENT_NAME, rightComponent);
+    return cfg;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public void execute(TupleWindow inputWindow) {
+    Map<K, KeyValue<V1, V2>> joinMap = new HashMap<>();
+    for (Tuple tuple : inputWindow.get()) {
+      if (tuple.getSourceComponent().equals(leftComponent)) {
+        KeyValue<K, V1> tup = (KeyValue<K, V1>) tuple.getValue(0);
+        if (tup.getKey() != null) {
+          addMapLeft(joinMap, tup);
+        }
+      } else {
+        KeyValue<K, V2> tup = (KeyValue<K, V2>) tuple.getValue(0);
+        if (tup.getKey() != null) {
+          addMapRight(joinMap, tup);
+        }
+      }
+    }
+    for (K key : joinMap.keySet()) {
+      KeyValue<V1, V2> val = joinMap.get(key);
+      switch (joinType) {
+        case INNER:
+          if (val.getKey() != null && val.getValue() != null) {
+            joinAndEmit(key, val);
+          }
+          break;
+        case LEFT:
+          if (val.getKey() != null) {
+            joinAndEmit(key, val);
+          }
+          break;
+        case OUTER:
+          joinAndEmit(key, val);
+          break;
+        default:
+          throw new RuntimeException("Unknown jointype");
```

Comment:
Best to include the unknown join type in the message.

Reply:
Added